pkg/controllers/subnetport/subnetport_controller.go

```diff
@@ -0,0 +1,290 @@
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/metrics"
+	v1 "k8s.io/api/core/v1"
+	apimachineryruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/controllers/common"
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/subnetport"
+)
+
+var (
+	log                     = logger.Log
+	MetricResTypeSubnetPort = common.MetricResTypeSubnetPort
+)
+
+// SubnetPortReconciler reconciles a SubnetPort object
+type SubnetPortReconciler struct {
+	client.Client
+	Scheme  *apimachineryruntime.Scheme
+	Service *subnetport.SubnetPortService
+}
+
+//+kubebuilder:rbac:groups=nsx.vmware.com,resources=subnetports,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=nsx.vmware.com,resources=subnetports/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=nsx.vmware.com,resources=subnetports/finalizers,verbs=update
+func (r *SubnetPortReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha1.SubnetPort{}
+	log.Info("reconciling subnetport CR", "subnetport", req.NamespacedName, "obj", obj)
+
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnetPort)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch subnetport CR", "req", req.NamespacedName)
+		return common.ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResTypeSubnetPort)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "subnetport", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return common.ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on subnetport CR", "subnetport", req.NamespacedName)
+		}
+
+		if len(obj.Spec.SubnetSet) > 0 && len(obj.Spec.Subnet) > 0 {
+			err := errors.New("subnet and subnetset should not be configured at the same time")
+			log.Error(err, "failed to get subnet/subnetset of the subnetport", "subnetport", req.NamespacedName)
+		}
+		if len(obj.Spec.SubnetSet) == 0 && len(obj.Spec.Subnet) == 0 {
```

Comment:
Should we make this judgement before adding finalizer to subnetport CR?

Reply:
good point, fixed.